### PR TITLE
[CALCITE-6257] StarRocks dialect implementation

### PR DESCRIPTION
**Pull request by YiwenWu**
Implementing StarRocks dialect in Calcite 
https://issues.apache.org/jira/browse/CALCITE-6257

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      break;
+    default:
+      super.unparseCall(writer, call, leftPrec, rightPrec);
+      break;
+    }
+  }
+
+  @Override public @Nullable SqlNode getCastSpec(RelDataType type) {
+    switch (type.getSqlTypeName()) {
+    case TIMESTAMP:
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java:119]: I am confused what's going on here.
If the Calcite TIMESTAMP type is called "DATETIME" in StarRocks, I am not sure whether this is the only case that needs changing. Can you write some tests that use TIMESTAMP?

File: core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import static org.apache.calcite.util.RelToSqlConverterUtil.unparseHiveTrim;
+
+/**
+ * A <code>SqlDialect</code> implementation for the StarRocks database.
+ */
+public class StarRocksSqlDialect extends SqlDialect {
```
[the remark this answers is not in the document]
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java:38]: As far as I know, the sql dialect of starrocks is compatible with mysql, but starrocks is an olap database that supports more data types and some special sql functions.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 3]
```diff
@@ -4506,13 +4600,16 @@ private void checkLiteral2(String expression, String expected) {
     String expectedPresto = "SELECT DATE_TRUNC('MINUTE', \"hire_date\")\n"
         + "FROM \"foodmart\".\"employee\"";
     String expectedFirebolt = expectedPostgresql;
+    String expectedStarRocks = "SELECT DATE_FORMAT(`hire_date`, '%Y-%m-%d %H:%i:00')\n"
```
[the remark this answers is not in the document]
Comment (YiwenWu) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:4603]: Indeed, this test case directly uses `MysqlSqlDialect#unparseFloor`,  MySQL/StarRocks Function DATE_FORMAT returns a string. [date_format docs](https://docs.starrocks.io/docs/2.1/sql-reference/sql-functions/date-time-functions/date_format/)

StarRocks also support function [date_trunc](https://docs.starrocks.io/docs/2.1/sql-reference/sql-functions/date-time-functions/date_trunc/),  replace with DATE_TRUNC function

File: core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java (hunk 1 above)
Comment (YiwenWu) [reply core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java:119]: Exactly,  Similar to MySQL, the Calcite TIMESTAMP type is called "DATETIME" in StarRocks.  This processing refers to `MysqlSqlDialect#getCastSpec`.

Comment (YiwenWu) [reply core/src/main/java/org/apache/calcite/sql/dialect/StarRocksSqlDialect.java:133]: Added additional TIMESTAMP test:
- `RelToSqlConverterTest#testCastToTimestamp`
- `RelToSqlConverterTest#testUnparseSqlIntervalQualifier`

**Comment by macroguo-ghy** [comment]
Please solve the conflicts first.

from [starrocks doc](https://docs.starrocks.io/docs/introduction/what_is_starrocks/), Starrocks say 'Compatible with MySQL', so maybe `StarrocksSqlDialect` can extend from `MysqlSqldialect`?

**Comment by YiwenWu** [comment]
> Please solve the conflicts first.
> 
> from [starrocks doc](https://docs.starrocks.io/docs/introduction/what_is_starrocks/), Starrocks say 'Compatible with MySQL', so maybe `StarrocksSqlDialect` can extend from `MysqlSqldialect`?

I'll try to handle it like this.

**Pull request closed without merging**